### PR TITLE
FW-3192 Add voice note feature in the chat-widget

### PR DESCRIPTION
**Pull request by Prochnost55**
### What do you want to achieve?
- Add voice note feature in the widget

### PR Checklist

- [x] I have tested it locally and all functionalities are working fine.
- [x] I have compared it with mocks and all design elements are the same.
- [x] I have tested it in IE Browser. 

### How was the code tested?

- init the chat widget with `voiceNote: true`
- start a conversation click on mic icon

### What new thing you came across while writing this code? 
-

### In case you fixed a bug then please describe the root cause of it? 
- 

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

File: webplugin/js/app/km-utils.js [hunk 1]
```diff
@@ -549,4 +549,43 @@ KommunicateUtils = {
                 .join('&')
         );
     },
+    IntervalTimer: function (callback, interval) {
```
Comment (devashishmamgain) [inline webplugin/js/app/km-utils.js]: camelCase

File: webplugin/js/app/km-utils.js [hunk 2]
```diff
@@ -549,4 +549,43 @@ KommunicateUtils = {
...
+        };
+        this.resume = function () {
+            if (state != 2) return;
+
+            state = 3;
+            window.setTimeout(this.timeoutCallback, remaining);
+        };
+
+        this.timeoutCallback = function () {
+            if (state != 3) return;
```
Comment (devashishmamgain) [inline webplugin/js/app/km-utils.js]: use { } for better readibility

File: webplugin/js/app/kommunicate-ui.js [hunk 3]
```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
             }
         );
     },
+    initRecorder: function(){
+        const LIVE_OUTPUT = false; // a feature to live output the recording voice to the speaker
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-ui.js:1813]: verify if const is working on all major browsers

File: webplugin/js/app/kommunicate-ui.js [hunk 4]
```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
...
+                        var minutesElapsed = Math.floor((timeElapsed / (60)) % 60);
+                        timeRemainingTimer.text(("0" + minutesElapsed).slice(-2) + ":" + ("0" + secondsElapsed).slice(-2));
+                        $("#wave-front-progressBar").width(percent+'%');
+                        if(REMAINING_TIME <=0){
+                            clearTimeout(playPauseInterval);
+                            pauseBtn.addClass("n-vis");
+                            playBtn.removeClass("n-vis");
+                            hasTimerExpired = true;
+                        }
+                    }else{
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-ui.js:1855]: code formatting

File: webplugin/js/app/kommunicate-ui.js [hunk 5]
```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
...
+                var timeElapsed = now - startTime;
+                var timeRemaining = endTime - now;
+                var percent = timeElapsed / MAX_RECORD_TIME * 100;
+                var secondsElapsed = Math.floor((timeElapsed / 1000) % 60);
+                var secondsRemaining = Math.floor((timeRemaining / 1000) % 60);
+                var minutesElapsed = Math.floor((timeElapsed / (1000 * 60)) % 60);
+                var minutesRemaining = Math.floor((timeRemaining / (1000 * 60)) % 60);
+                timeElapsedTimer.text(("0" + minutesElapsed).slice(-2) + ":" + ("0" + secondsElapsed).slice(-2));
+                timeRemainingTimer.text(("0" + minutesRemaining).slice(-2) + ":" + ("0" + secondsRemaining).slice(-2));
+                $("#wave-front-progressBar").width(percent+'%');
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-ui.js:1915]: what's $ here? do we need to use $applozic?

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate-ui.js:1915]: same at other places

File: webplugin/js/app/kommunicate-ui.js [hunk 6]
```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
...
+            Kommunicate.typingAreaService.hideRecorder();
+            !permissionDenied && Fr.voice.stop();
+            clearInterval(recorderInterval);
+            clearInterval(playPauseInterval);
+            recorderInterval = null;
+            audioBlob = null;
+            wavAudioDuration = null;
+            recorderAudio.setAttribute('src', '');
+            isTimerPaused = false;
+            hasTimerExpired = true;
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-ui.js:1976]: are these related? is it possible isTimerPaused and hasTimerExpired are both true or both false at the same time? or they both are always going to be opposite of each other?

File: webplugin/js/app/kommunicate-ui.js [hunk 7]
```diff
@@ -1803,4 +1803,220 @@ KommunicateUI = {
...
+            params = {};
+            $("#wave-front-progressBar").width('0%');
+            timeElapsedTimer.text("00:00");
+            timeRemainingTimer.text("02:00");
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate-ui.js:1980]: is 2 mins hard coded?

File: webplugin/css/app/mck-sidebox-1.0.css [hunk 8]
```diff
@@ -6372,4 +6347,79 @@ body.accesibility :focus {
 .mck-msg-left:last-child [data-hidepostformsubmit="true"] {
 	display: block;
 	visibility: visible;
+}
+
+/* ----------------- VoiceNote feature ------------- */
```
Comment (devashishmamgain) [inline webplugin/css/app/mck-sidebox-1.0.css:6352]: can we move voice related code both css and js to a separate file? these 2 files are becoming larger and larger with new code, will be difficult to work on these

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate-ui.js (hunk 7 above)
Comment (Prochnost55) [reply webplugin/js/app/kommunicate-ui.js:1980]: yes

File: webplugin/js/app/km-utils.js (hunk 1 above)
Comment (Prochnost55) [reply webplugin/js/app/km-utils.js]: this was not required so removed this

File: webplugin/js/app/km-utils.js (hunk 2 above)
Comment (Prochnost55) [reply webplugin/js/app/km-utils.js]: It is also removed

File: webplugin/js/app/kommunicate-ui.js (hunk 3 above)
Comment (Prochnost55) [reply webplugin/js/app/kommunicate-ui.js:1813]: Screenshot 2022-01-05 at 4 56 54 PM

For IE and opera mini their is partial support.

File: webplugin/js/app/kommunicate-ui.js (hunk 5 above)
Comment (Prochnost55) [reply webplugin/js/app/kommunicate-ui.js:1915]: forgot to replace $ with $applozic. Corrected now.

File: webplugin/js/app/kommunicate-ui.js (hunk 6 above)
Comment (Prochnost55) [reply webplugin/js/app/kommunicate-ui.js:1976]: if the timer is running then `isTimerPaused = false` and also `hasTimerExpired = false`.
if the timer is paused then `isTimerPaused = true` and also `hasTimerExpired = false`.
if the timer is cleared then `isTimerPaused = false` and also `hasTimerExpired = true`.

File: webplugin/css/app/mck-sidebox-1.0.css (hunk 8 above)
Comment (Prochnost55) [reply webplugin/css/app/mck-sidebox-1.0.css:6352]: Moved to different files

Comment (devashishmamgain) [reply webplugin/css/app/mck-sidebox-1.0.css]: saw the file km-voice-note.css but mck-sidebox-1.0.css is still showing new css for voice

**Review by devashishmamgain — DISMISSED** [review]

File: webplugin/js/app/media/media-service.js [hunk 9]
```diff
@@ -137,4 +137,222 @@ Kommunicate.mediaService = {
...
+                        var minutesElapsed = Math.floor((timeElapsed / (60)) % 60);
+                        timeRemainingTimer.text(("0" + minutesElapsed).slice(-2) + ":" + ("0" + secondsElapsed).slice(-2));
+                        $applozic("#wave-front-progressBar").width(percent+'%');
+                        if(REMAINING_TIME <=0){
+                            clearTimeout(playPauseInterval);
+                            playPausetimerState = TIMER_STATE.EXPIRED;
+                            pauseBtn.addClass("n-vis");
+                            playBtn.removeClass("n-vis");            
+                        };
+                    }else if(playPausetimerState == TIMER_STATE.PAUSED){
```
Comment (devashishmamgain) [inline webplugin/js/app/media/media-service.js:185]: code formatting

**Review by devashishmamgain — APPROVED** [review]

**Review by gargshantnu — APPROVED** [review]
